core/src/main/scala/org/apache/spark/rdd/RDD.scala

```diff
@@ -1233,6 +1233,21 @@ abstract class RDD[T: ClassTag](
           (i, iter) => iter.map((i % curNumPartitions, _))
         }.foldByKey(zeroValue, new HashPartitioner(curNumPartitions))(cleanCombOp).values
       }
+      if (conf.get(ENABLE_EXECUTOR_TREE_AGGREGATE) && partiallyAggregated.partitions.length > 1) {
+        // define a new partitioner that results in only 1 partition
+        val constantPartitioner = new Partitioner {
+          override def numPartitions: Int = 1
+
+          override def getPartition(key: Any): Int = 0
+        }
+        // map the partially aggregated rdd into a key-value rdd
+        // do the computation in the single executor with one partition
+        // get the new RDD[U]
+        partiallyAggregated = partiallyAggregated
+          .map(v => (0.toByte, v))
+          .foldByKey(zeroValue, constantPartitioner)(cleanCombOp)
+          .values
+      }
       val copiedZeroValue = Utils.clone(zeroValue, sc.env.closureSerializer.newInstance())
```

Comment:
wouldn't this submit another job after the `foldByKey` again on the `fold` eventually to just return the only value?

Reply:
no it would not submit another job, because there is only one partition. spark would do the fold internal the partition (we only have 1 in this case), and return. 
As described on Apache Spark's website the fold definition states:

```
"Aggregate the elements of each partition, and then the results for all the partitions, using a given associative function and a neutral "zero value"."
```
Thus fold operates by folding first each element of partition and then the result of partition
it is partition dependent